### DISCUSSION TITLE
Different result in different themes

### DISCUSSION THREAD

**Discussion by JouriM2** [AvaloniaUI/Avalonia#15890] [Q&A] [answered]
Here is code sample: https://github.com/JouriM2/AvDropDownBug
Here I summarize all problems I cant solve by myself.
This sample contains several separated problems I found.
Can somebody point out what am I doing wrong?

1. I have a bindings to `CheckBox` `Context` and `IsChecked` like this:
```HTML
<CheckBox Content="{Binding ThemeName}" IsChecked="{Binding IsFluentTheme}"/>
```
and in code like this:
```C#
public partial class MainWindowViewModel : ObservableValidator, INotifyPropertyChanged {
    string ThemeName {
        get => $"Theme is \"{App.CurrentTheme}\"";
    }
    bool IsFluentTheme {
        get => App.CurrentTheme == App.CatalogTheme.Fluent;
        set => App.SetCatalogThemes( value ? App.CatalogTheme.Fluent : App.CatalogTheme.Simple );
    }
```

If VisualStudio xaml preview it works as supposed

![изображение](https://github.com/AvaloniaUI/Avalonia/assets/114450272/3213b73c-6bb4-4856-afef-d9edaa993cf9)

but then i start application (release or debug) these bindings not working: getter and setter for both properties not called.

2. I have a button which switch application theme (code for this are taken from Avalonia controls demo).
The theme is switched, but then window recreated, new windows position shifted to right and bottom for several pixels even I set it in "App.axaml.cs" using:
```C#
                if ( oldWindow is not null ) 
                    newWindow.Position = oldWindow.Position;
```
3. I need to display data in popup list of `ComboBox` aligned to the width of this control. I use `Grid` to align items but its not fully working in `Fluent` theme:

![изображение](https://github.com/AvaloniaUI/Avalonia/assets/114450272/3f7fe695-a69a-4813-8cc0-1a74289b6fbd)

and completelly not working in `Simple` theme:

![изображение](https://github.com/AvaloniaUI/Avalonia/assets/114450272/746e04d3-8894-4011-9393-41b250be67ce)

Same problem with simple `StackPanel` in `Simple` theme (first ComboBox in sample) - I cant find a way to make `StackPanel` use whole list line:

![изображение](https://github.com/AvaloniaUI/Avalonia/assets/114450272/9f3a0f1d-9f7f-4faa-83a2-f2dd5a438d27)

**Comment by stevemonaco** [comment]
Theme switching is not really a thing out-of-the-box, despite being technically possible. Each theme defines its own sizes, margins, padding, alignments, etc which are part of the look-and-feel. As such, you can get drastically incompatible results as you're currently seeing.

You need to read the `ControlTheme`s to customize them. That's non-negotiable. See the two `ControlTheme`s for `ComboBoxItem`:

https://github.com/AvaloniaUI/Avalonia/blob/7db46af7c86be30e7585283fa8cea9222c39200c/src/Avalonia.Themes.Fluent/Controls/ComboBoxItem.xaml#L18-L24

https://github.com/AvaloniaUI/Avalonia/blob/9be975b0e4874bc5d1cad4872615db40959a2430/src/Avalonia.Themes.Simple/Controls/ComboBoxItem.xaml#L4-L11

You can see that Fluent has `Stretch` and Simple has `Left` for `HorizontalContentAlignment`. You could create a `ComboBoxItem` selector so both match, but you're going to run into numerous cases of things like this.

**Reply by JouriM2** [reply]
Thanks, now I know what to fix it.
Actually I dont need theme switching in my app and add it to the sample just to illustrate difference.
I was trying to use simple theme as it looks more familiar to standalone app users.
I expect key control properties be the same in all themes.
Is this a "bug" in Avalonia and need to be fixed or I need to redefine this property in every app with simple  theme?

How about other questions?
F.i. is it possible to use grid for drop down items and get columns of the same sizes in each line?
I think I know my problem: I define independent grids for each line, so I need some place to define one grid for thole dropdown part and only set `Grid.Column` in each line item.
How to do this?

**Reply by stevemonaco** [reply, marked as answer]
AFAIK, Simple Theme is meant to closely mimic WPF's default theme. The same non-stretching behavior is there for other collection items like `TreeViewItem` and probably the others (`ListBoxItem`, `ComboBoxItem`, etc). So I don't think it's a bug, but I also didn't author the theme.

Check `Grid.IsSharedSizeScope` (probably set this on your `ComboBox`) and then `SharedSizeGroup` for the `Grid` created by each item (via `<DataTemplate>` I assume). [Relevant docs](https://docs.avaloniaui.net/docs/reference/controls/grid/sharedsizegroup)

**Reply by JouriM2** [reply]
Thanks again!
Just may be it will be usefull for somebody.

**To make `ComboBox` items fill whole dropdown lines in SimpleTheme**

```HTML
		<ComboBox ...>
			<ComboBox.Styles>
				<Style Selector="ComboBoxItem">
					<Setter Property="HorizontalContentAlignment" Value="Stretch" />
				</Style>
			</ComboBox.Styles>
			<ComboBox.ItemTemplate>
				<DataTemplate ...>
					<Panel Background="Aqua" HorizontalAlignment="Stretch">
						<TextBlock HorizontalAlignment="Left" .../>
						<TextBlock HorizontalAlignment="Center" .../>
						<TextBlock HorizontalAlignment="Right" .../>
					</Panel>
				</DataTemplate>
			</ComboBox.ItemTemplate>
		</ComboBox>
```
Now Panel fill whole line and items placed as expected:
![изображение](https://github.com/AvaloniaUI/Avalonia/assets/114450272/759a45c3-8a69-4500-82e2-041bd89b1f41)

**To use Grid in ComboBox\ListBox items with same columns widths**

1. Apply the same `HorizontalContentAlignment` to `ComboBoxItem`
2. Use columns groups in data template
```HTML
		<ComboBox Grid.IsSharedSizeScope="True" ...">
				<DataTemplate ...>
					<Grid Background="Aqua">
						<Grid.ColumnDefinitions>
							<ColumnDefinition SharedSizeGroup="A" Width="*" />
							<ColumnDefinition SharedSizeGroup="B" Width="*" />
							<ColumnDefinition SharedSizeGroup="C" Width="Auto" />
						</Grid.ColumnDefinitions>
						<TextBlock Grid.Column="0" .../>
						<TextBlock Grid.Column="1" .../>
						<TextBlock Grid.Column="2" .../>
					</Grid>
				</DataTemplate>
		</ComboBox>
```
Now all columns are the same in whole grid:
![изображение](https://github.com/AvaloniaUI/Avalonia/assets/114450272/e322ac84-074f-4633-ac88-a6927ff3adb5)

**Reply by JouriM2** [reply]
The last (but most annoying) problem is non-working properties.
May be you take a look at this too?

**Reply by stevemonaco** [reply]
Might be better to create a new issue. In general, a `ControlTheme` can support as many or few dependency properties on the target control as it wants.
